src/features/import-map.js

```diff
@@ -4,46 +4,90 @@
 
-var importMap, importMapPromise;
+var importMapPromise, fetchPromises = {};
+
+// Start downloading the import maps as soon as possible
 if (typeof document !== 'undefined') {
-  const scripts = document.getElementsByTagName('script');
-  for (let i = 0; i < scripts.length; i++) {
-    const script = scripts[i];
-    if (script.type !== 'systemjs-importmap')
-      continue;
-
-    if (!script.src) {
-      importMap = parseImportMap(JSON.parse(script.innerHTML), pageBaseUrl);
-    }
-    else {
-      importMapPromise = fetch(script.src)
+  const scriptsToDownload = document.querySelectorAll('script[type="systemjs-importmap"][src]');
+
+  for (let i = 0; i < scriptsToDownload.length; i++) {
+    const script = scriptsToDownload[i]
+    fetchPromises[script.src] = fetch(script.src)
       .then(function (res) {
         return res.json();
       })
-      .then(function (json) {
-        importMap = parseImportMap(json, script.src);
-      });
+      .then(function (data) {
+        return parseImportMap(data, script.src);
+      })
+  }
+}
+
+function retrieveImportMaps() {
+  if (typeof document !== 'undefined') {
+    // We must apply/merge the import maps in the order that they appear in the dom
+    const scripts = document.querySelectorAll('script[type="systemjs-importmap"]');
+    let pendingMaps = []
+
+    for (let i = 0; i < scripts.length; i++) {
+      const script = scripts[i];
+      if (!script.src) {
+        pendingMaps.push(parseImportMap(JSON.parse(script.innerHTML), pageBaseUrl));
+      }
+      else if (fetchPromises[script.src]) {
+        pendingMaps.push(fetchPromises[script.src])
+      } else {
+        pendingMaps.push(
+          fetch(script.src)
+          .then(function (res) {
+            return res.json();
+          })
+          .then(function (data) {
+            return parseImportMap(data, script.src);
+          })
+        );
+      }
     }
-    break;
+
+    // free the memory for this object -- we don't need it anymore after the first System.import happens
+    fetchPromises = null
+
+    const initialImportMap = { imports: {}, scopes: {} };
+
+    return Promise.all(pendingMaps).then(function(importMaps) {
+      return importMaps.reduce(function(finalMap, importMap) {
+        return mergeImportMap(finalMap, importMap);
+      }, initialImportMap);
+    });
   }
 }
 
-importMap = importMap || { imports: {}, scopes: {} };
+export function mergeImportMap(originalMap, newMap) {
+  for (let i in newMap.imports) {
+    originalMap.imports[i] = newMap.imports[i];
+  }
+
+  for (let i in newMap.scopes) {
+    originalMap.scopes[i] = newMap.scopes[i];
+  }
+
+  originalMap.baseUrl = newMap.baseUrl;
+
+  return originalMap;
+}
 
 systemJSPrototype.resolve = function (id, parentUrl) {
   parentUrl = parentUrl || pageBaseUrl;
 
-  if (importMapPromise)
-    return importMapPromise
-    .then(function () {
+  if (!importMapPromise) {
+    importMapPromise = retrieveImportMaps();
+  }
```

Comment:
Instead of `retrieveImportMaps`, if this promise is just always populated on startup we can just remove this conditional codepath entirely.

Reply:
Reading through [this part of the sec](https://github.com/WICG/import-maps/blob/master/spec.md#when-import-maps-can-be-encountered), I'm not sure if what you said is true. It seems like you have to keep track of whether this is the first time that a `<script type="module">` (or `System.import()`) is encountered so that you know to ignore import maps created after "acquiring import maps" is set to false.

Reply:
Ah, I was just recently discussing with Domenic how we might permit dynamic import maps, and the current spec there seems to exclude any concept of dynamic loading due to this state being quite difficult to work with.

What I would like to see the spec do is the following:
1. Import maps get processed in DOM order, with module scripts also getting processed in DOM order. As import maps are processed they update the `importMapPromise`.
2. Any script processing or resolve operation will wait on the current `importMapPromise`. This way a `<script type="module">` will only wait on the import maps loaded before it.

So that's sort of what I was describing. We need to have more discussion on the spec side, but I feel getting a base that supports dynamic import maps simply is important.

Reply:
Admittedly we should really interleave the importMapPromises with the moduleInstantiatePromises to get this right... since an import map should block the next module script from loading, etc.

Reply:
Ok, lets start in SystemJS by not supporting dynamic import maps, and just follow the spec exactly. That will simplify the algorithm from having to do the above. Will re-review based on this approach.

Reply:
@guybedford I am confused because you are saying let's follow the spec exactly by not supporting dynamic imports, but [the spec does support dynamic imports](https://github.com/WICG/import-maps#dynamic-import-map-example).